coordinator/subscriber.go

```diff
@@ -0,0 +1,373 @@
+		fn(dbi)
+	}
+}
+
+func (s *SubscriberManager) UpdateWriters() {
+	s.WalkDatabases(func(dbi *meta.DatabaseInfo) {
+		if _, ok := s.writers[dbi.Name]; !ok {
+			s.writers[dbi.Name] = make(map[string][]SubscriberWriter)
+		}
+		dbi.WalkRetentionPolicy(func(rpi *meta.RetentionPolicyInfo) {
+			changed := false
+			s.lock.RLock()
+			writers, ok := s.writers[dbi.Name][rpi.Name]
+			s.lock.RUnlock()
+			if !ok {
+				writers = make([]SubscriberWriter, 0, len(rpi.Subscriptions))
+				changed = true
+			}
+			// record origin subscription names in a set
+			originSubs := make(map[string]bool)
+			for _, w := range writers {
+				originSubs[w.Name()] = true
+			}
+			// add new subscriptions
+			for _, sub := range rpi.Subscriptions {
+				if _, ok := originSubs[sub.Name]; !ok {
+					writer, err := s.NewSubscriberWriter(dbi.Name, rpi.Name, sub.Name, sub.Mode, sub.Destinations)
+					if err != nil {
+						s.Logger.Info("fail to create subscriber", zap.String("db", dbi.Name), zap.String("rp", rpi.Name), zap.String("sub", sub.Name),
+							zap.Strings("dest", sub.Destinations))
+					} else {
+						writers = append(writers, writer)
+						writer.Start(s.config.WriteConcurrency, s.config.WriteBufferSize)
+						s.Logger.Info("add new subscriber writer", zap.String("db", dbi.Name), zap.String("rp", rpi.Name), zap.String("sub", sub.Name),
+							zap.Strings("dest", sub.Destinations))
+						changed = true
+					}
+				}
+				// remove all appeared subscription from the set
+				// then rest names are of the subscriptions that should be removed
+				delete(originSubs, sub.Name)
+			}
+			// if there is no subscriptions to remove,
+			// just continue
+			if len(originSubs) == 0 {
+				if changed {
+					s.lock.Lock()
+					s.writers[dbi.Name][rpi.Name] = writers
+					s.lock.Unlock()
+				}
+				return
+			}
+			position := 0
+			length := len(writers)
+			for i := 0; i < length; i++ {
+				if _, ok := originSubs[writers[i].Name()]; !ok {
+					writers[position] = writers[i]
+					position++
+				} else {
+					writers[i].Stop()
+					s.Logger.Info("remove subscriber writer", zap.String("db", dbi.Name), zap.String("rp", rpi.Name), zap.String("sub", writers[i].Name()))
+				}
+			}
+			s.lock.Lock()
+			s.writers[dbi.Name][rpi.Name] = writers[0:position]
+			s.lock.Unlock()
+		})
+	})
+}
+
+func (s *SubscriberManager) Send(db, rp string, lineProtocal []byte) {
+	if rp == "" {
+		dbi, err := s.client.Database(db)
+		if err != nil {
+			s.Logger.Error("unknown database", zap.String("db", db))
+		} else {
+			rp = dbi.DefaultRetentionPolicy
+		}
+	}
+
+	if writer, ok := s.writers[db][rp]; ok {
+		for _, w := range writer {
+			w.Write(lineProtocal)
+		}
+	}
+}
+
+func (s *SubscriberManager) StopAllWriters() {
+	for _, db := range s.writers {
+		for _, rp := range db {
+			for _, writer := range rp {
+				writer.Stop()
+			}
+		}
+	}
+}
+
+func (s *SubscriberManager) Update() {
+	for {
+		ch := s.client.(*metaclient.Client).WaitForDataChanged()
```

Comment:
Is this conflict with the function waitForIndex? waitForIndex wait chis channel.

Reply:
Actually there is no conflict, `pollForUpdates` will notify all the channels when metadata is changed, 
see https://github.com/openGemini/openGemini/blob/main/lib/metaclient/meta_client.go#L2399